Config-flow for DLNA-DMR integration

chishm opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Breaking change

The DLNA-DMR integration migrated to configuration via the UI. Configuring DLNA-DMR via YAML configuration has been deprecated and will be removed in a future Home Assistant release.

Your existing YAML configuration is automatically imported on upgrade to this release; and thus can be safely removed from your YAML configuration after upgrading.

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds config-flow configuration for the `dlna_dmr` component, and deprecates YAML configuration. It is a big change, and in support of this, it also adds unit tests and strict mypy checking.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

The primary aim of this pull request is to add config-flow support and SSDP auto-discovery to the `dlna_dmr` component. It already supported SSDP, but disabled-by-default, using the legacy `discovery` component; now it's done using `DlnaDmrFlowHandler.async_step_ssdp`.

I've also improved the handling of unavailable devices, re-using the `ssdp` component for device availability notifications instead of polling for missing devices on the network. This might lead to a slightly increased delay for when devices become available within HA when turned on (from 10 seconds `async_update` polling interval to 1 minute SSDP search). If a device is not available when HA starts up then the entity will still be created, in an unavailable state.

To aid development, I've added strict typing and unit tests.

I've also added @StevenLooman to the code owners, as he is the original author of the component, but the code owners was blank.

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/19087

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

chishm committed:
Modernize dlna_dmr component: configflow, test, types

* Support config-flow with ssdp discovery
* Add unit tests
* Enforce strict typing
* Gracefully handle network devices (dis)appearing

chishm committed:
Fix Aiohttp mock response headers type to match actual response class

chishm committed:
Fixes from code review

chishm commented on homeassistant/components/dlna_dmr/manifest.json:
```diff
@@ -1,9 +1,21 @@
+      "deviceType": "urn:schemas-upnp-org:device:MediaRenderer:3"
+    }
+  ],
+  "codeowners": ["@StevenLooman"],
```
Done. I wasn't sure on the etiquette, and @StevenLooman has done most of the hard work here and with async_upnp_client.

chishm commented on homeassistant/components/dlna_dmr/data.py:
```diff
@@ -0,0 +1,94 @@
+        self.upnp_factory = UpnpFactory(self.requester, non_strict=True)
+        self.event_notifiers = {}
+
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.async_shutdown)
```
I'm not sure how to achieve that, since the data is shared by all `ConfigEntry`s (it's the integration's global data in `hass.data`). Is there an integration-wide `async_shutdown` (complement to `async_setup`) that could be used?

bdraco replied:
When the async entry unload is called check to see how many config entries are still loaded with the same domain, and if it's the last one, call your unloader

chishm replied:
Thanks for the tip. I've gone with a slightly different resource management method, releasing the event_notifier when the device is disconnected, and cancelling the shutdown listener when there's no more event_notifiers.

chishm commented:
@bdraco Thank you for the code review. I'll figure out what is going on with the import attempts and come back with further comments/changes.

chishm committed:
Fixes from code review

chishm commented:
@bdraco I've looked into the failed import attempts and I think they've failed because the device is off when HA starts (I've added extra logging to confirm). They should be picked up via SSDP when they're next turned on, but any extra settings (`listen_port`, `name`,  and `callback_url_override`) from the YAML config won't be imported. Is this limitation acceptable, or should it make more effort to contact the device (e.g. polling for it)?

chishm committed:
Import device config in flow if unavailable at hass start

chishm commented:
I've updated the config-flow to apply settings from configuration.yml to added devices if they weren't already migrated. This should catch the case of a device being off when HA starts and turned on after.

StevenLooman reviewed (approved):
Great work @chishm!

I hadn't seen the review request, sorry about that. If this happens again, feel free to send me a message on discord or any other channel.

The integration depends less on the URL, great! This will simplify things by not having to assign a static IP to a device, helping many people.

In https://github.com/home-assistant/core/pull/55540 I'm trying to use SSDP search responses and SSDP advertisements to better monitor devices. For example, devices can send a SSDP advertisement `BYEBYE` when it shuts down, or `ALIVE` when it is back online. I think this can be used to know that a device is no longer available, is available again, respectively. Although there are probably plenty of devices which don't send advertisements at all... and the integration has to rely on its own implementation to monitor the device.

There still is some manual configuration wrt eventing callback. I suspect this is only needed in multi-network installations. Can the `network` component supply any required information here such that this manual configuration is no longer needed?

This PR looks good to me. Just a few general questions.

chishm commented on homeassistant/components/dlna_dmr/media_player.py:
```diff
@@ -24,209 +25,301 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the DlnaDmrEntity from a config entry."""
+    del hass  # Unused
```
Mostly to satisfy linting. HA's pylint configuration disables the `unused-argument` warning, but I've gotten into the habit of using `del` to indicate that it's intentional.

chishm commented on homeassistant/components/dlna_dmr/media_player.py:
```diff
@@ -24,209 +25,301 @@
+                "name": self._device.name,
+                # Connection is based on the root device's UDN, which is
+                # currently equivalent to our UDN (embedded devices aren't
+                # supported by async_upnp_client)
```
The one device I've come across with an embedded (i.e., not at the root level of the UPnP description XML) DMR device is Sonos speakers. Since they're already covered by the `sonos` integration, I don't think there's a pressing need to support it. But if there ever is another device like that, I have thoughts about how to support it.

StevenLooman replied:
Feel free to create an issue in the `async_upnp_client` repo with your ideas. No need to implement it directly, saving the thoughts is already helpful.

chishm replied:
Done: https://github.com/StevenLooman/async_upnp_client/issues/80

StevenLooman replied:
Thanks!

chishm commented on homeassistant/components/dlna_dmr/config_flow.py:
```diff
@@ -0,0 +1,331 @@
+        return DlnaDmrOptionsFlowHandler(config_entry)
+
+    async def async_step_user(self, user_input: FlowInput = None) -> FlowResult:
+        """Handle a flow initialized by the user: manual URL entry.
```
Indeed, most users should be supported by the automatic SSDP discovery step. This, like the `listen_port` and `callback_url_override` are meant for the case that SSDP is broken for some reason (e.g. as you said multi-network installations, or when HA is behind a NAT or firewall).

I originally had code to let users pick from a list of discovered devices, then realised that the discovery was done in the exact same way as the usual ssdp step, so it didn't actually provide any benefits. Superfluous code still needs to be tested, so I removed the code instead.

chishm commented:
@StevenLooman Thank you for the code review.

The [SSDP PR](https://github.com/home-assistant/core/pull/55540) looks like something I've wanted too, and I am happy to submit a follow-up PR to the `dlna_dmr` component to use it when both of these PRs get merged. I think calling `DlnaDmrEntity._device_disconnect` should suffice for a BYEBYE, and `DlnaDmrEntity._device_connect` for an ALIVE. The UPDATE case is more complicated, but can probably be handled by disconnect followed by connect (when the location changes), like in `DlnaDmrEntity.async_config_update_listener`.

StevenLooman commented:
I think this also solves #48190 (unique IDs)

bdraco commented:
The original version I tested was blocking startup as observed in https://github.com/home-assistant/core/pull/55540

Once https://github.com/home-assistant/core/pull/55540 merges and this gets rebased, I'll come back and test again.

bdraco commented:
@chishm This will need a few tweaks for the breaking change in https://github.com/home-assistant/core/pull/55540.  Let me know when its ready and I'll retest to ensure it doesn't block startup

chishm committed:
Support SSDP advertisements

chishm commented:
@bdraco Thanks for testing with the latest ssdp changes. I've modified this PR to support the SSDP advertisements feature, and it should now be quickly responsive to devices being turned on and off.

Unfortunately there's a small issue with `ssdp:byebye` messages not getting through that I'm currently debugging. I'll push a fix here when it's resolved.

bdraco reviewed (changes requested):
I'm seeing the following tracebacks on restart

```
2021-09-13 20:08:32 ERROR (MainThread) [homeassistant.components.ssdp] Failed to callback info: {'CACHE-CONTROL': 'max-age=1800', 'DATE': 'Mon, 13 Sep 2021 20:08:31 GMT', 'ssdp_ext': '', 'BOOTID.UPNP.ORG': 'FF760004-CB26-45EB-B561-D58DFF760004', 'ssdp_server': 'Linux/3.0.8', 'ssdp_st': 'urn:schemas-upnp-org:device:MediaRenderer:1', 'ssdp_usn': 'uuid:FF760004-CB26-45EB-B561-D58DFF760004::urn:schemas-upnp-org:device:MediaRenderer:1', '_location_original': 'http://192.168.106.96:49152/description.xml', 'ssdp_location': 'http://192.168.106.96:49152/description.xml', '_timestamp': datetime.datetime(2021, 9, 13, 20, 8, 32, 362301), '_host': '192.168.106.96', '_port': 58341, '_udn': 'uuid:FF760004-CB26-45EB-B561-D58DFF760004', 'X_DLNADOC': 'DMR-1.50', 'deviceType': 'urn:schemas-upnp-org:device:MediaRenderer:1', 'friendlyName': 'Office New Onelink', 'manufacturer': 'First Alert', 'manufacturerURL': 'https://firstalert.ca', 'modelDescription': 'Onelink Safe Sound', 'modelName': 'Onelink Safe Sound', 'modelURL': 'https://firstalert.ca', 'UDN': 'uuid:FF760004-CB26-45EB-B561-D58DFF760004', 'modelNumber': 'V01-Oct 20 2020', 'serialNumber': '00001', 'ssidName': 'Onelink Safe Sound 5B9C', 'uuid': 'FF760004CB2645EBB561D58D', 'X_QPlay_SoftwareCapability': 'QPlay:2', 'iconList': {'icon': [{'mimetype': 'image/png', 'width': '48', 'height': '48', 'depth': '24', 'url': '/upnp/grender-48x48.png'}, {'mimetype': 'image/png', 'width': '120', 'height': '120', 'depth': '24', 'url': '/upnp/grender-120x120.png'}, {'mimetype': 'image/jpeg', 'width': '48', 'height': '48', 'depth': '24', 'url': '/upnp/grender-48x48.jpg'}, {'mimetype': 'image/jpeg', 'width': '120', 'height': '120', 'depth': '24', 'url': '/upnp/grender-120x120.jpg'}]}, 'serviceList': None}
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/ssdp/__init__.py", line 142, in _async_process_callbacks
    await callback(discovery_info, ssdp_change)
  File "/usr/src/homeassistant/homeassistant/components/dlna_dmr/media_player.py", line 195, in async_ssdp_callback
    bootid = int(bootid_str, 10)
ValueError: invalid literal for int() with base 10: 'FF760004-CB26-45EB-B561-D58DFF760004'
2021-09-13 20:08:33 ERROR (MainThread) [homeassistant.components.ssdp] Failed to callback info: {'CACHE-CONTROL': 'max-age=1800', 'DATE': 'Mon, 13 Sep 2021 20:08:31 GMT', 'ssdp_ext': '', 'BOOTID.UPNP.ORG': 'FF760004-46E1-80E1-BBE3-9D0AFF760004', 'ssdp_server': 'Linux/3.0.8', 'ssdp_st': 'urn:schemas-upnp-org:device:MediaRenderer:1', 'ssdp_usn': 'uuid:FF760004-46E1-80E1-BBE3-9D0AFF760004::urn:schemas-upnp-org:device:MediaRenderer:1', '_location_original': 'http://192.168.106.52:49152/description.xml', 'ssdp_location': 'http://192.168.106.52:49152/description.xml', '_timestamp': datetime.datetime(2021, 9, 13, 20, 8, 32, 883268), '_host': '192.168.106.52', '_port': 52429, '_udn': 'uuid:FF760004-46E1-80E1-BBE3-9D0AFF760004', 'X_DLNADOC': 'DMR-1.50', 'deviceType': 'urn:schemas-upnp-org:device:MediaRenderer:1', 'friendlyName': 'Onelink 13F0', 'manufacturer': 'First Alert', 'manufacturerURL': 'https://firstalert.ca', 'modelDescription': 'Onelink Safe Sound', 'modelName': 'Onelink Safe Sound', 'modelURL': 'https://firstalert.ca', 'UDN': 'uuid:FF760004-46E1-80E1-BBE3-9D0AFF760004', 'modelNumber': 'V01-Oct 20 2020', 'serialNumber': '00001', 'ssidName': 'Onelink Safe Sound 13F0', 'uuid': 'FF76000446E180E1BBE39D0A', 'X_QPlay_SoftwareCapability': 'QPlay:2', 'iconList': {'icon': [{'mimetype': 'image/png', 'width': '48', 'height': '48', 'depth': '24', 'url': '/upnp/grender-48x48.png'}, {'mimetype': 'image/png', 'width': '120', 'height': '120', 'depth': '24', 'url': '/upnp/grender-120x120.png'}, {'mimetype': 'image/jpeg', 'width': '48', 'height': '48', 'depth': '24', 'url': '/upnp/grender-48x48.jpg'}, {'mimetype': 'image/jpeg', 'width': '120', 'height': '120', 'depth': '24', 'url': '/upnp/grender-120x120.jpg'}]}, 'serviceList': None}
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/ssdp/__init__.py", line 142, in _async_process_callbacks
    await callback(discovery_info, ssdp_change)
  File "/usr/src/homeassistant/homeassistant/components/dlna_dmr/media_player.py", line 195, in async_ssdp_callback
    bootid = int(bootid_str, 10)
ValueError: invalid literal for int() with base 10: 'FF760004-46E1-80E1-BBE3-9D0AFF760004'
2021-09-13 20:08:33 ERROR (MainThread) [homeassistant.components.ssdp] Failed to callback info: {'CACHE-CONTROL': 'max-age=1800', 'DATE': 'Mon, 13 Sep 2021 20:08:31 GMT', 'ssdp_ext': '', 'BOOTID.UPNP.ORG': 'FF760004-8EF2-D0F6-1A9D-F2AFFF760004', 'ssdp_server': 'Linux/3.0.8', 'ssdp_st': 'urn:schemas-upnp-org:device:MediaRenderer:1', 'ssdp_usn': 'uuid:FF760004-8EF2-D0F6-1A9D-F2AFFF760004::urn:schemas-upnp-org:device:MediaRenderer:1', '_location_original': 'http://192.168.107.249:49152/description.xml', 'ssdp_location': 'http://192.168.107.249:49152/description.xml', '_timestamp': datetime.datetime(2021, 9, 13, 20, 8, 33, 756366), '_host': '192.168.107.249', '_port': 41818, '_udn': 'uuid:FF760004-8EF2-D0F6-1A9D-F2AFFF760004', 'X_DLNADOC': 'DMR-1.50', 'deviceType': 'urn:schemas-upnp-org:device:MediaRenderer:1', 'friendlyName': 'Guest Room Onelink', 'manufacturer': 'First Alert', 'manufacturerURL': 'https://firstalert.ca', 'modelDescription': 'Onelink Safe Sound', 'modelName': 'Onelink Safe Sound', 'modelURL': 'https://firstalert.ca', 'UDN': 'uuid:FF760004-8EF2-D0F6-1A9D-F2AFFF760004', 'modelNumber': 'V01-Oct 20 2020', 'serialNumber': '00001', 'ssidName': 'Onelink Safe Sound 9D1C', 'uuid': 'FF7600048EF2D0F61A9DF2AF', 'X_QPlay_SoftwareCapability': 'QPlay:2', 'iconList': {'icon': [{'mimetype': 'image/png', 'width': '48', 'height': '48', 'depth': '24', 'url': '/upnp/grender-48x48.png'}, {'mimetype': 'image/png', 'width': '120', 'height': '120', 'depth': '24', 'url': '/upnp/grender-120x120.png'}, {'mimetype': 'image/jpeg', 'width': '48', 'height': '48', 'depth': '24', 'url': '/upnp/grender-48x48.jpg'}, {'mimetype': 'image/jpeg', 'width': '120', 'height': '120', 'depth': '24', 'url': '/upnp/grender-120x120.jpg'}]}, 'serviceList': None}
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/ssdp/__init__.py", line 142, in _async_process_callbacks
    await callback(discovery_info, ssdp_change)
  File "/usr/src/homeassistant/homeassistant/components/dlna_dmr/media_player.py", line 195, in async_ssdp_callback
    bootid = int(bootid_str, 10)
ValueError: invalid literal for int() with base 10: 'FF760004-8EF2-D0F6-1A9D-F2AFFF760004'
2021-09-13 20:08:33 ERROR (MainThread) [homeassistant.components.ssdp] Failed to callback info: {'CACHE-CONTROL': 'max-age=1800', 'DATE': 'Mon, 13 Sep 2021 20:08:31 GMT', 'ssdp_ext': '', 'BOOTID.UPNP.ORG': 'FF760004-381B-8B5A-602D-646AFF760004', 'ssdp_server': 'Linux/3.0.8', 'ssdp_st': 'urn:schemas-upnp-org:device:MediaRenderer:1', 'ssdp_usn': 'uuid:FF760004-381B-8B5A-602D-646AFF760004::urn:schemas-upnp-org:device:MediaRenderer:1', '_location_original': 'http://192.168.107.51:49152/description.xml', 'ssdp_location': 'http://192.168.107.51:49152/description.xml', '_timestamp': datetime.datetime(2021, 9, 13, 20, 8, 33, 885158), '_host': '192.168.107.51', '_port': 40099, '_udn': 'uuid:FF760004-381B-8B5A-602D-646AFF760004', 'X_DLNADOC': 'DMR-1.50', 'deviceType': 'urn:schemas-upnp-org:device:MediaRenderer:1', 'friendlyName': 'Master Bed Onelink', 'manufacturer': 'First Alert', 'manufacturerURL': 'https://firstalert.ca', 'modelDescription': 'Onelink Safe Sound', 'modelName': 'Onelink Safe Sound', 'modelURL': 'https://firstalert.ca', 'UDN': 'uuid:FF760004-381B-8B5A-602D-646AFF760004', 'modelNumber': 'V01-Oct 20 2020', 'serialNumber': '00001', 'ssidName': 'Onelink Safe Sound D3B4', 'uuid': 'FF760004381B8B5A602D646A', 'X_QPlay_SoftwareCapability': 'QPlay:2', 'iconList': {'icon': [{'mimetype': 'image/png', 'width': '48', 'height': '48', 'depth': '24', 'url': '/upnp/grender-48x48.png'}, {'mimetype': 'image/png', 'width': '120', 'height': '120', 'depth': '24', 'url': '/upnp/grender-120x120.png'}, {'mimetype': 'image/jpeg', 'width': '48', 'height': '48', 'depth': '24', 'url': '/upnp/grender-48x48.jpg'}, {'mimetype': 'image/jpeg', 'width': '120', 'height': '120', 'depth': '24', 'url': '/upnp/grender-120x120.jpg'}]}, 'serviceList': None}
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/ssdp/__init__.py", line 142, in _async_process_callbacks
    await callback(discovery_info, ssdp_change)
  File "/usr/src/homeassistant/homeassistant/components/dlna_dmr/media_player.py", line 195, in async_ssdp_callback
    bootid = int(bootid_str, 10)
ValueError: invalid literal for int() with base 10: 'FF760004-381B-8B5A-602D-646AFF760004'
```

chishm commented:
> I'm seeing the following tracebacks on restart
> 
> ```
> 2021-09-13 20:08:32 ERROR (MainThread) [homeassistant.components.ssdp] Failed to callback info: {'CACHE-CONTROL': 'max-age=1800', 'DATE': 'Mon, 13 Sep 2021 20:08:31 GMT', 'ssdp_ext': '', 'BOOTID.UPNP.ORG': 'FF760004-CB26-45EB-B561-D58DFF760004', 'ssdp_server': 'Linux/3.0.8', 'ssdp_st': 'urn:schemas-upnp-org:device:MediaRenderer:1', 'ssdp_usn': 'uuid:FF760004-CB26-45EB-B561-D58DFF760004::urn:schemas-upnp-org:device:MediaRenderer:1', '_location_original': 'http://192.168.106.96:49152/description.xml', 'ssdp_location': 'http://192.168.106.96:49152/description.xml', '_timestamp': datetime.datetime(2021, 9, 13, 20, 8, 32, 362301), '_host': '192.168.106.96', '_port': 58341, '_udn': 'uuid:FF760004-CB26-45EB-B561-D58DFF760004', 'X_DLNADOC': 'DMR-1.50', 'deviceType': 'urn:schemas-upnp-org:device:MediaRenderer:1', 'friendlyName': 'Office New Onelink', 'manufacturer': 'First Alert', 'manufacturerURL': 'https://firstalert.ca', 'modelDescription': 'Onelink Safe Sound', 'modelName': 'Onelink Safe Sound', 'modelURL': 'https://firstalert.ca', 'UDN': 'uuid:FF760004-CB26-45EB-B561-D58DFF760004', 'modelNumber': 'V01-Oct 20 2020', 'serialNumber': '00001', 'ssidName': 'Onelink Safe Sound 5B9C', 'uuid': 'FF760004CB2645EBB561D58D', 'X_QPlay_SoftwareCapability': 'QPlay:2', 'iconList': {'icon': [{'mimetype': 'image/png', 'width': '48', 'height': '48', 'depth': '24', 'url': '/upnp/grender-48x48.png'}, {'mimetype': 'image/png', 'width': '120', 'height': '120', 'depth': '24', 'url': '/upnp/grender-120x120.png'}, {'mimetype': 'image/jpeg', 'width': '48', 'height': '48', 'depth': '24', 'url': '/upnp/grender-48x48.jpg'}, {'mimetype': 'image/jpeg', 'width': '120', 'height': '120', 'depth': '24', 'url': '/upnp/grender-120x120.jpg'}]}, 'serviceList': None}
> Traceback (most recent call last):
>   File "/usr/src/homeassistant/homeassistant/components/ssdp/__init__.py", line 142, in _async_process_callbacks
>     await callback(discovery_info, ssdp_change)
>   File "/usr/src/homeassistant/homeassistant/components/dlna_dmr/media_player.py", line 195, in async_ssdp_callback
>     bootid = int(bootid_str, 10)
> ValueError: invalid literal for int() with base 10: 'FF760004-CB26-45EB-B561-D58DFF760004'
> ```

That looks like the device doesn't conform to the UPnP spec. I'll make the parsing a bit more tolerant.

chishm commented:
This depends on https://github.com/StevenLooman/async_upnp_client/pull/84 to fix the `ssdp:byebye` problem I had.

chishm committed:
Ignore bad BOOTID, fix ssdp:byebye handling

chishm committed:
Only listen for events on interface connected to device

chishm commented:
Updated and working, all except for the ssdp:byebye problem that should be fixed when async_upnp_client is updated.

StevenLooman commented:
Merged the fix @chishm + optimizations by @bdraco. Do we want a release of `async_upnp_client`, or are there more optimizations coming up @bdraco ?

bdraco commented:
> Merged the fix @chishm + optimizations by @bdraco. Do we want a release of `async_upnp_client`, or are there more optimizations coming up @bdraco ?

Nothing more coming for at least today.  I'm out of local production cases I feel comfortable upgrading.  When I'm back in Texas later this month I'll be able to test on a few more places so there might be more at that time.

StevenLooman commented:
Released `async_upnp_client==0.21.3`

StevenLooman commented:
Bumped async_upnp_client in this PR, hope you don't mind @chishm.

bdraco commented on homeassistant/components/dlna_dmr/media_player.py:
```diff
@@ -24,209 +27,346 @@
-        vol.Optional(CONF_CALLBACK_URL_OVERRIDE): cv.url,
-    }
+PLATFORM_SCHEMA = vol.All(
+    cv.deprecated(CONF_URL),
```
Should the whole domain be marked as deprecated?

StevenLooman replied:
Has my preference. By the looks of it all options can be set via the OptionsFlow.

chishm replied:
The `dlna_dmr` yaml config is done as a platform within the `media_player` domain, and gets used like so:
```
media_player:
  - platform: dlna_dmr
    url: http://fake/fake.xml
```

I'm not sure how to mark this as deprecated, since `cv.deprecated` only applies to keys, not key/value combos.

bdraco replied:
I think we have done a logger message with level of warning that it's deprecated advising removal of the yaml config in the past for this case.

bdraco replied:
Example #50312

chishm replied:
Thanks for the example, and done.

chishm committed:
Release all listeners when entities are removed

chishm committed:
Warn about deprecated dlna_dmr configuration

chishm committed:
Use sublogger for dlna_dmr.config_flow for easier filtering

chishm committed:
Tests for dlna_dmr.data module

bdraco commented on tests/components/dlna_dmr/test_data.py:
```diff
@@ -0,0 +1,70 @@
+    assert get_domain_data(hass) is domain_data
+
+
+async def test_event_notifier(hass: HomeAssistant) -> None:
```
Generally we try not to interact with the integration details and patch the external libraries in tests:
https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco commented on tests/components/dlna_dmr/test_init.py:
```diff
@@ -0,0 +1,66 @@
+        ]
+    }
+
+    with patch.object(
```
https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

bdraco commented on tests/components/dlna_dmr/test_media_player.py:
```diff
@@ -0,0 +1,871 @@
+) -> None:
+    """Test a DlnaDmrEntity with a connected DmrDevice."""
+    # The physical device should not have been contacted yet
+    assert device_requests_mock.call_count == 0
```
https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations

Setup the integration and test by patching the external package and assert the entity state via the core state machine `hass.states`.

chishm commented:
> Bumped async_upnp_client in this PR, hope you don't mind @chishm.

Not at all. Thanks for merging the PR, it's all working now when tested against my devices.

bdraco commented:
https://github.com/StevenLooman/async_upnp_client/blob/development/async_upnp_client/client_factory.py#L60

Maybe we should limit `async_create_device` to this list?
```
  "ssdp": [
    {
      "deviceType": "urn:schemas-upnp-org:device:MediaRenderer:1"
    },
    {
      "deviceType": "urn:schemas-upnp-org:device:MediaRenderer:2"
    },
    {
      "deviceType": "urn:schemas-upnp-org:device:MediaRenderer:3"
    }
  ],
```

Here is a Denon AMP that has 4 devices (`urn:schemas-upnp-org:device:MediaRenderer:1` included) that gets rejected
[aios_device.xml.zip](https://github.com/home-assistant/core/files/7174998/aios_device.xml.zip)

chishm commented:
> Here is a Denon AMP that has 4 devices (`urn:schemas-upnp-org:device:MediaRenderer:1` included) that gets rejected with `Device is not a Digital Media Receiver` even though I can use it in a few DLNA clients

I had the same problem with a Sonos speaker. It's due to the issue I flagged here https://github.com/StevenLooman/async_upnp_client/issues/80 . In the Sonos case I was able to prevent its discovery by restricting the ssdp filter to the list you mentioned, and the device would never send SSDP search responses for those device types specifically. I'm guessing that's not the case for the Denon AMP.

We could remove the check for `device.device_type` in [`DlnaDmrFlowHandler._async_connect`](https://github.com/home-assistant/core/pull/55267/files#diff-f45651f0903da654503dcc706fa65a875299bf7b825ff890f4ef8e5e4a7a0addR258), which would allow it through but potentially cause issues as outlined in the linked issue.

bdraco commented:
> StevenLooman/async_upnp_client#80

I agree that approach #3 seems like the a good path forward.

StevenLooman commented on homeassistant/components/dlna_dmr/strings.json:
```diff
@@ -0,0 +1,44 @@
+      "discovery_error": "Failed to discover a matching DLNA device",
+      "incomplete_config": "Configuration is missing a required variable",
+      "non_unique_id": "Multiple devices found with the same unique ID",
+      "not_dmr": "Device is not a Digital Media Receiver"
```
Technically, DMR stands for Digital Media Renderer, not receiver.

StevenLooman replied:
In fact, home assistant becomes a DMC (Digital Media Controller), so perhaps the integration is named wrong? Though a bit late to change things...

chishm replied:
Whoops, fixed now. I think the integration is still named correctly, since most integrations are named for what they control and not what they are.

StevenLooman commented:
You are right about the embedded devices and Sonos. What is weird however, is that I am able to call a service using the CLI:
```shell
$ upnp-client call-action http://192.168.178.19:1400/xml/device_description.xml AVTransport/GetCurrentTransportActions InstanceID=0
{"timestamp": 1632085416.4645689, "service_id": "urn:upnp-org:serviceId:AVTransport", "service_type": "urn:schemas-upnp-org:service:AVTransport:1", "action": "GetCurrentTransportActions", "in_parameters": {"InstanceID": 0}, "out_parameters": {"Actions": "Set, Stop, Play"}}
```

Device description parts:
```xml
<?xml version="1.0" encoding="utf-8"?>
<root xmlns="urn:schemas-upnp-org:device-1-0">
  <specVersion>
    <major>1</major>
    <minor>0</minor>
  </specVersion>
  <device>
    <deviceType>urn:schemas-upnp-org:device:ZonePlayer:1</deviceType>
    <friendlyName>192.168.178.19 - SYMFONISK Bookshelf</friendlyName>
...
    <deviceList>
...
      <device>
        <deviceType>urn:schemas-upnp-org:device:MediaRenderer:1</deviceType>
        <friendlyName>Woonkamer - SYMFONISK Bookshelf Media Renderer</friendlyName>
...
        <serviceList>
...
          <service>
            <serviceType>urn:schemas-upnp-org:service:AVTransport:1</serviceType>
            <serviceId>urn:upnp-org:serviceId:AVTransport</serviceId>
            <controlURL>/MediaRenderer/AVTransport/Control</controlURL>
            <eventSubURL>/MediaRenderer/AVTransport/Event</eventSubURL>
            <SCPDURL>/xml/AVTransport1.xml</SCPDURL>
          </service>
...
        </serviceList>
...
      </device>
    </deviceList>
  </device>
</root>
```

Feel free to start work on https://github.com/StevenLooman/async_upnp_client/issues/80. There are a few other things I want to finish, home assistant related and non-home assistant related.

StevenLooman commented:
http://upnp.org/specs/av/UPnP-av-MediaRenderer-v3-Device.pdf mentions:
> A  MediaRenderer  device  can  be  either  a  Root  device  or  can  be  Embedded  in  another  UPnP  device  (MediaRenderer  or  other).  A  MediaRenderer device (Root or Embedded) can in turn contain other standard or non-standard Embedded UPnP devices. 

When initially writing `async_upnp_client`, my media renderer devices were devices which only had a single (root-)device, thus supporting embedded devices was never needed.

StevenLooman commented:
You know what, I'll try to pick https://github.com/StevenLooman/async_upnp_client/issues/80 up later this week. Can't make any promises though.

chishm commented:
> You know what, I'll try to pick [StevenLooman/async_upnp_client#80](https://github.com/StevenLooman/async_upnp_client/issues/80) up later this week. Can't make any promises though.

I'll finish rewrtiting the unit tests here then I'll see how you're going with that and have a go if you haven't already.

chishm committed:
Rewrite DMR tests for HA style

chishm committed:
Fix DMR strings: "Digital Media *Renderer*"

chishm committed:
Update DMR entity state and device info when changed

chishm committed:
Replace deprecated async_upnp_client State with TransportState

chishm committed:
supported_features are dynamic, based on current device state

chishm committed:
Cleanup fully when subscription fails

chishm committed:
Log warnings when device connection fails unexpectedly

chishm commented:
I've rewritten the tests to follow the guidelines in [Writing tests for integrations](https://developers.home-assistant.io/docs/development_testing/#writing-tests-for-integrations). Coverage is at 99% across the module, with only 2 lines missed in `dlna_dmr/media_player.py`.

Would it be possible to get a quality rating? I'm aiming for Platinum, but will settle for Gold if that's what @StevenLooman is comfortable with (Platinum requires approval by code owners for every PR).

chishm committed:
Set PARALLEL_UPDATES to unlimited

chishm committed:
Fix spelling

StevenLooman commented:
> Would it be possible to get a quality rating? I'm aiming for Platinum, but will settle for Gold if that's what @StevenLooman is comfortable with (Platinum requires approval by code owners for every PR).

Wow, great work! To be honest I've never tried for a quality rating for the `upnp` component and have no experience with it whatsoever. I'm leaving it up to others what should be done here.

bdraco commented:
> Would it be possible to get a quality rating? I'm aiming for Platinum, but will settle for Gold if that's what @StevenLooman is comfortable with (Platinum requires approval by code owners for every PR).

Let's do the quality rating in a new PR after this merges as this PR is already quite large.  Hopefully it will just be one or two small tweaks (or nothing besides setting it)

bdraco commented:
Also tried entering the url manually in the config flow
http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml

no dice

bdraco commented on homeassistant/components/dlna_dmr/media_player.py:
```diff
@@ -24,298 +27,479 @@
+        new_port = entry.options.get(CONF_LISTEN_PORT) or 0
+        new_callback_url = entry.options.get(CONF_CALLBACK_URL_OVERRIDE)
+
+        if (
+            new_port != self._event_addr.port
+            or new_callback_url != self._event_addr.callback_url
+        ):
```
Reverse the conditional and return then outdent the block below

bdraco commented on homeassistant/components/dlna_dmr/media_player.py:
```diff
@@ -24,298 +27,479 @@
+                _LOGGER.debug("Error while subscribing during device connect: %s", err)
+                raise
+
+        if (
+            self.registry_entry
+            and self.registry_entry.config_entry_id
+            and not self.registry_entry.device_id
+        ):
```
reverse the condition, return, and outdent the block below

chishm committed:
Fixes from code review

chishm committed:
Simplify has & can checks to just can, which includes has

chishm committed:
Treat transitioning state as playing (not idle) to reduce UI jerking

chishm commented:
> Wow, great work! To be honest I've never tried for a quality rating for the upnp component and have no experience with it whatsoever. I'm leaving it up to others what should be done here.

Thank you :smile:  I'll shoot for Platinum when it's the right time to do so.

> Let's do the quality rating in a new PR after this merges as this PR is already quite large. Hopefully it will just be one or two small tweaks (or nothing besides setting it)

Sounds like a good plan.

> Also tried entering the url manually in the config flow
> no dice

Yeah, it uses the same validation pathway as the SSDP discovery flow. Once https://github.com/StevenLooman/async_upnp_client/pull/86 is merged I'll be able to get this working.

StevenLooman committed:
Test if device is usable

StevenLooman commented:
I've updated both `async-upnp-client` (https://github.com/StevenLooman/async_upnp_client/pull/86) and this branch.

Some observations:

Symfonisk devices are not discovered, but can be added manually/through url. Strange, as a search results in a `urn:schemas-upnp-org:device:MediaRenderer:1` response.

Added 3 devices (2x Symfonisk, 1x TV.) Removing 1x Symfonisk is ok, from then on removing a device results in this error:
```
2021-09-23 20:42:54 ERROR (MainThread) [homeassistant.config_entries] Error unloading entry 192.168.178.24 - SYMFONISK Bookshelf for media_player
Traceback (most recent call last):
  File "/home/steven/src/python/home-assistant/homeassistant/config_entries.py", line 443, in async_unload
    result = await component.async_unload_entry(hass, self)  # type: ignore
  File "/home/steven/src/python/home-assistant/homeassistant/components/media_player/__init__.py", line 372, in async_unload_entry
    return await component.async_unload_entry(entry)
  File "/home/steven/src/python/home-assistant/homeassistant/helpers/entity_component.py", line 183, in async_unload_entry
    await platform.async_reset()
  File "/home/steven/src/python/home-assistant/homeassistant/helpers/entity_platform.py", line 601, in async_reset
    await asyncio.gather(*tasks)
  File "/home/steven/src/python/home-assistant/homeassistant/helpers/entity.py", line 722, in async_remove
    await self.async_will_remove_from_hass()
  File "/home/steven/src/python/home-assistant/homeassistant/components/dlna_dmr/media_player.py", line 193, in async_will_remove_from_hass
    callback()
  File "/home/steven/src/python/home-assistant/homeassistant/components/ssdp/__init__.py", line 214, in _async_remove_callback
    self._callbacks.remove(callback_entry)
ValueError: list.remove(x): x not in list
```

Samsung TV is discovered. Adding it works as expected. Can be removed. After removal (no hass restart), it can no longer automatically be added.

Haven't tested playing any media.

`upnp` seemed to break, not sure what is happening there. I'll test it a bit more, but please also test it yourself if you have the ability.

StevenLooman commented:
> `upnp` seemed to break, not sure what is happening there. I'll test it a bit more, but please also test it yourself if you have the ability.

Seems to be working again after a restart.

> Symfonisk devices are not discovered, but can be added manually/through url. Strange, as a search results in a urn:schemas-upnp-org:device:MediaRenderer:1 response, seen using upnp-client search.

Sonos doesn't show up as the in `dlna_dmr/manifest.json` uses on `deviceType`:
```
  "ssdp": [
    {
      "deviceType": "urn:schemas-upnp-org:device:MediaRenderer:1"
    },
    {
      "deviceType": "urn:schemas-upnp-org:device:MediaRenderer:2"
    },
    {
      "deviceType": "urn:schemas-upnp-org:device:MediaRenderer:3"
    }
  ],
```

When the media renderer device is an embedded device, the actual `deviceType` in `info_with_desc` (in `homeassistant.components.ssdp.Scanner._ssdp_listener_callback`) is for the root device, not the embedded device. It'll probably work when in `manifest.json`, `deviceType` is replaced by `st` (in case of searches) or `nt` (in case of advertisements.) I think `st`/search is more reliable.

chishm commented:
> Symfonisk devices are not discovered, but can be added manually/through url. Strange, as a search results in a urn:schemas-upnp-org:device:MediaRenderer:1 response, seen using upnp-client search.

> When the media renderer device is an embedded device, the actual `deviceType` in `info_with_desc` (in `homeassistant.components.ssdp.Scanner._ssdp_listener_callback`) is for the root device, not the embedded device, thus the domain is not matched. It'll probably work when in `manifest.json`, `deviceType` is replaced by `st` (in case of searches) or `nt` (in case of advertisements.) I think `st`/search is more reliable. Or, perhaps there are other solutions (either on the hass side or on the `async-upnp-client` side?)

Indeed. I'd purposely done that to avoid the device not working when auto-discovered, especially since SONOS has its own integration.

But when embedded devices are working I'll switch it to using the st/nt header.

> Added 3 devices (2x Symfonisk, 1x TV.) Removing 1x Symfonisk is ok, from then on removing a device results in this error:

Ugh, I just spotted [my error](https://github.com/home-assistant/core/pull/55267/files#diff-1a430fcc1ceb0948e838d4aca53a83eb08589a2a10b0305bdb61456d68a70217R142). Auto-initialised class-level attributes are a bad idea for lists.

chishm committed:
Handle ssdp:update messages properly

chishm committed:
Fix _remove_ssdp_callbacks being shared by all DlnaDmrEntity instances

chishm committed:
Fix tests for transitioning state

chishm committed:
Mock DmrDevice.is_profile_device (added to support embedded devices)

chishm committed:
Use ST & NT SSDP headers to find DMR devices, not deviceType

The deviceType is extracted from the device's description XML, and will not
be what we want when dealing with embedded devices.

chishm committed:
Use UDN from SSDP headers, not device description, as unique_id

The SSDP headers have the UDN of the embedded device that we're interested
in, whereas the device description (`ATTR_UPNP_UDN`) field will always be
for the root device.

chishm commented:
Discovery is now done based on Service Type (ST, as sent for M-SEARCH responses) and Notification Type (NT, as sent for ssdp:alive advertisements). The unique_id is now derived from the embedded device's UDN, not the root device, so this might show a double discovery if you've been testing your own local branch.

This PR now has a dependency on https://github.com/StevenLooman/async_upnp_client/pull/86 and won't pass CI until that is merged. I'll update this PR when it is.

chishm commented:
> Samsung TV is discovered. Adding it works as expected. Can be removed. After removal (no hass restart), it can no longer automatically be added.

It might reappear after HA does another search, or after it advertises itself again. Try waiting for a minute (HA search period), or try turning the TV off and on again.

chishm committed:
Fix DMR string English localization

chishm committed:
Test config flow with UDN from SSDP headers

StevenLooman commented:
@chishm Since there is another PR (https://github.com/StevenLooman/async_upnp_client/pull/87), do you think https://github.com/StevenLooman/async_upnp_client/pull/86 can be merged, or are there other bugs to be found/squahed? If

StevenLooman commented:
I see your comments on the PR, thank you for those. Will fix those today.

StevenLooman committed:
Bump async-upnp-client==0.22.1, fix flake8 error

bdraco commented:
Assuming https://github.com/home-assistant/core/pull/56591 passes the CI, I'll merge it, and this one should pass the test again

bdraco committed:
fix test for remapping

bdraco commented:
<img width="946" alt="Screen Shot 2021-09-26 at 5 31 13 PM" src="https://user-images.githubusercontent.com/663432/134826305-ae719b74-1aae-4848-bd4c-a1aeac15e545.png">

Its finding the embedded devices now

bdraco commented:
```
2021-09-26 17:31:36 INFO (MainThread) [homeassistant.helpers.entity_registry] Registered new media_player.dlna_dmr entity: media_player.marantz_nr1711
2021-09-26 17:31:36 ERROR (MainThread) [async_upnp_client.profiles.profile] Failed subscribing to service: 
```

bdraco commented on tests/components/dlna_dmr/conftest.py:
```diff
@@ -0,0 +1,136 @@
+        yield
+
+
+@pytest.fixture(autouse=True)
```
This looks similar to https://github.com/home-assistant/core/pull/56431

We should clean this up in a new PR once both merge.

chishm committed:
DMR HA Device connections based on root and embedded UDN

chishm committed:
DmrDevice's UpnpDevice is now named profile_device

chishm committed:
Use device type from SSDP headers, not device description

chishm committed:
Mark dlna_dmr constants as Final

bdraco reviewed (approved):
This looks good.

There may be another adjustment needed to `async-upnp-client` per https://github.com/home-assistant/core/pull/55267#issuecomment-927383576 or it may be the device I was testing with doesn't conform to the spec because others seem ok.

chishm commented:
Thank you again for the detailed review, @bdraco  and @StevenLooman.

I'm not entirely sure what went wrong with that Denon speaker. I'll improve the debug messages for `UpnpError` (which at the moment rendered blank in the message `[async_upnp_client.profiles.profile] Failed subscribing to service:  `) to hopefully get some more useful information.

I've been testing this locally and there's a few minor things for me to fix, namely listening for SSDP alives from embedded devices (at the moment a Sonos speaker will never become "available" after it goes offline). I'll get that working today and push the fixes.

bdraco commented:
I marked it as a draft to make sure it doesn't get merged before your tweaks get pushed since there are usually more reviewers to cleanup the PR queue a bit more before Wednesdays' beta.

chishm committed:
Use embedded device UDN and type for unique ID when connected via URL

chishm committed:
More informative connection error messages

chishm committed:
Also match SSDP messages on NT headers

The NT header is to ssdp:alive messages what ST is to M-SEARCH responses.

chishm commented:
This is now working fully for me. But it has a dependency on https://github.com/StevenLooman/async_upnp_client/pull/89 , so is not ready for merge just yet.

StevenLooman committed:
Bump async-upnp-client==0.22.2

StevenLooman commented:
Thank you for the fixes to async-upnp-client @chishm. Released 0.22.2 + bumped version in this PR.

bdraco commented:
rebased to fix the conflict

bdraco committed:
fix merge

bdraco commented:
Error is coming through now on the denon 

`2021-09-27 12:03:18 ERROR (MainThread) [async_upnp_client.profiles.profile] Failed subscribing to service: UpnpConnectionError('Server disconnected')`

bdraco commented:
So the device is closing the connection which is technically allowed by the http spec at any time but aiohttp will throw ServerDisconnectedError

bdraco commented:
When I add the retry on ServerDisconnectedError everything works with the denon device

bdraco commented:
Looks like we are hammering the device with requests

```
2021-09-27 12:37:41 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: SUBSCRIBE http://192.168.209.173:1400/MediaRenderer/RenderingControl/Event {'NT': 'upnp:event', 'TIMEOUT': 'Second-540', 'HOST': '192.168.209.173:1400', 'CALLBACK': '<http://192.168.213.87:64807/notify>'} None
2021-09-27 12:37:41 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: SUBSCRIBE http://192.168.209.173:1400/MediaRenderer/AVTransport/Event {'NT': 'upnp:event', 'TIMEOUT': 'Second-540', 'HOST': '192.168.209.173:1400', 'CALLBACK': '<http://192.168.213.87:64807/notify>'} None
2021-09-27 12:37:41 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:41 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:41 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:41 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
2021-09-27 12:37:42 WARNING (MainThread) [async_upnp_client.aiohttp] HTTP R 0: GET http://192.168.213.223:60006/upnp/desc/aios_device/aios_device.xml None None
```

bdraco reviewed (approved):
For the errors noted above, I created PRs https://github.com/StevenLooman/async_upnp_client/pull/92 and https://github.com/StevenLooman/async_upnp_client/pull/93 

Everything is working AFAICT with these fixes.  This shouldn't hold up this PR though since these are existing problems with the library and not specifically related to this integration.

bdraco commented:
@chishm Any more tweaks ?

StevenLooman committed:
Bump async-upnp-client==0.22.3

bdraco reviewed (approved):
Retest passed 👍

chishm commented:
Wow, a lot has happened while I slept! Thanks bdraco and StevenLooman for the final fixes to get this over the line, and for getting it merged :smile: 

I'm currently working on a follow-up PR to implement more of the `MediaPlayerEntity` attributes, which shouldn't be such a significant change.

bdraco commented:
> Wow, a lot has happened while I slept! Thanks bdraco and StevenLooman for the final fixes to get this over the line, and for getting it merged :smile: 
> 
> 
> 
> I'm currently working on a follow-up PR to implement more of the `MediaPlayerEntity` attributes, which shouldn't be such a significant change.

If you keep it small, we will try to get it in before beta on Wednesday￼.  It would be nice for it to be a little more feature complete since it will be much more visible now.￼

chishm commented:
Here we go: https://github.com/home-assistant/core/pull/56728

There's more things I want to add (support for repeat and shuffle, enqueuing tracks, setting metadata when calling play_media), but they won't be ready for Wednesday.

MartinHjelmare commented on homeassistant/components/dlna_dmr/media_player.py:
```diff
@@ -24,298 +27,499 @@
+                )
+
+        # Device could have been de/re-connected, state probably changed
+        self.schedule_update_ha_state()
```
We're not allowed to use the sync api inside the event loop. Please use `self.async_write_ha_state` instead.

MartinHjelmare commented on homeassistant/components/dlna_dmr/__init__.py:
```diff
@@ -1 +1,56 @@
+    if MEDIA_PLAYER_DOMAIN not in config:
+        return True
+
+    for entry_config in config[MEDIA_PLAYER_DOMAIN]:
```
This way of importing platform config is not robust. It doesn't catch all accepted ways of writing the YAML for platforms. We should import the platform config from `async_setup_platform` in the media_player platform instead.

MartinHjelmare commented on homeassistant/components/dlna_dmr/config_flow.py:
```diff
@@ -0,0 +1,340 @@
+                    "Entry %s not imported: %s", import_data[CONF_URL], err.args[0]
+                )
+                # Store the config to apply if the device is added later
+                get_domain_data(self.hass).unmigrated_config[location] = import_data
```
This is not a good pattern. We want the config flow to be isolated from the rest of the integration. The flow should just return the config flow result.

MartinHjelmare commented on homeassistant/components/dlna_dmr/media_player.py:
```diff
@@ -24,298 +27,499 @@
+        self, hass: HomeAssistant, entry: config_entries.ConfigEntry
+    ) -> None:
+        """Handle options update by modifying self in-place."""
+        del hass  # Unused
```
Why do this?

MartinHjelmare commented on homeassistant/components/dlna_dmr/media_player.py:
```diff
@@ -24,298 +27,499 @@
-            hass, server_host, server_port, requester, callback_url_override
+
+    entry.async_on_unload(
+        entry.add_update_listener(entity.async_config_update_listener)
```
All entity state update callbacks must be registered in `async_added_to_hass` and unregistered on entity remove.

MartinHjelmare commented on homeassistant/components/dlna_dmr/media_player.py:
```diff
@@ -325,81 +529,90 @@ async def async_play_media(self, media_type, media_id, **kwargs):
+        ):
             return STATE_PAUSED
+        if self._device.transport_state == TransportState.VENDOR_DEFINED:
+            return STATE_UNKNOWN
```
Don't use `STATE_UNKNOWN` directly. We use `None` to represent unknown state. It will be converted in the entity base class.

bdraco replied:
This one was already addressed in #56728

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
